server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java

```diff
@@ -0,0 +1,196 @@
+
+import org.apache.james.dnsservice.api.DNSService;
+import org.apache.james.dnsservice.api.InMemoryDNSService;
+import org.apache.james.mailets.configuration.CommonProcessors;
+import org.apache.james.mailets.configuration.MailetConfiguration;
+import org.apache.james.mailets.configuration.MailetContainer;
+import org.apache.james.mailets.configuration.ProcessorConfiguration;
+import org.apache.james.modules.protocols.ImapGuiceProbe;
+import org.apache.james.modules.protocols.SmtpGuiceProbe;
+import org.apache.james.transport.mailets.RemoteDelivery;
+import org.apache.james.transport.matchers.All;
+import org.apache.james.util.docker.DockerContainer;
+import org.apache.james.utils.DataProbeImpl;
+import org.apache.james.utils.IMAPMessageReader;
+import org.apache.james.utils.SMTPMessageSender;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import io.restassured.builder.RequestSpecBuilder;
+import io.restassured.http.ContentType;
+import io.restassured.specification.RequestSpecification;
+
+public class RemoteDeliveryErrorTest {
+    private static final String ANOTHER_DOMAIN = "other.com";
+
+    private static final String FROM = "from@" + DEFAULT_DOMAIN;
+    private static final String RECIPIENT = "touser@" + ANOTHER_DOMAIN;
+    private static final String ALWAYS_421_RCPT_BEHAVIOR = "[{" +
+        "  \"condition\": {\"operator\":\"matchAll\"}," +
+        "  \"response\": {\"code\":421, \"message\":\"mock response\", \"rejected\":true}," +
+        "  \"command\": \"RCPT TO\"" +
+        "}]";
+    private static final String ALWAYS_421_FROM_BEHAVIOR = "[{" +
+        "  \"condition\": {\"operator\":\"matchAll\"}," +
+        "  \"response\": {\"code\":421, \"message\":\"mock response\", \"rejected\":true}," +
+        "  \"command\": \"MAIL FROM\"" +
+        "}]";
+    private static final String ALWAYS_421_DATA_BEHAVIOR = "[{" +
+        "  \"condition\": {\"operator\":\"matchAll\"}," +
+        "  \"response\": {\"code\":421, \"message\":\"mock response\", \"rejected\":true}," +
+        "  \"command\": \"DATA\"" +
+        "}]";
+    private static final String BOUNCE_MESSAGE = "Hi. This is the James mail server at localhost.\n" +
+        "I'm afraid I wasn't able to deliver your message to the following addresses.\n" +
+        "This is a permanent error; I've given up. Sorry it didn't work out. Below\n" +
+        "I include the list of recipients and the reason why I was unable to deliver\n" +
+        "your message.";
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+    @Rule
+    public IMAPMessageReader imapMessageReader = new IMAPMessageReader();
+    @Rule
+    public SMTPMessageSender messageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
+    @ClassRule
+    public static DockerContainer mockSmtp = DockerContainer.fromName("chibenwa/mock-smtp-server")
```

Comment:
how about linagora/mock-smtp-server ?

Reply:
For sure, but the image is not approved yet hence I temporarily use my repository.